Add utility functions for testing shrinkers

jonathanknowles opened the pull request:
## Issue Number

ADP-1538

## Summary

This PR adds a set of utility functions for testing shrinkers:

- `genShrinkSequence :: (a -> [a]) -> a -> Gen [a]`
    Generates a random sequence of progressively shrunken values from a given starting value and shrinking function. Terminates with a value that cannot be shrunk further with the given shrinking function.
- `shrinkWhile :: (a -> Bool) -> (a -> [a]) -> a -> Maybe a`
    Repeatedly applies a shrinking function to a value while a condition holds. Terminates with the smallest shrunken value that cannot be shrunk further with the given shrinking function.
- `shrinkWhileSteps :: (a -> Bool) -> (a -> [a]) -> a -> [a]`
    Repeatedly applies a shrinking function to a value while a condition holds, returning all the intermediate steps (lazily).
- `shrinkSpace :: Ord a => (a -> [a]) -> a -> Set a`
    Returns the set of all values to which a given value can be shrunk, for a given shrinking function.

jonathanknowles committed:
Add testing functions `shrinkWhile` and `shrinkWhileSteps`.

jonathanknowles committed:
Add testing function `shrinkSpace`.

jonathanknowles committed:
Add testing function `genShrinkSequence`.

jonathanknowles committed:
Add dependency on the `safe` library.

sevanspowell commented on lib/test-utils/test/Test/QuickCheck/ExtraSpec.hs:
```diff
@@ -590,6 +621,88 @@ prop_selectMapEntries_union m0 (Positive (Small i)) =
+    tests =
+        [ UnitTestData
+            { params = (1024, (>= 0))
+            , result = [0]
```
Shouldn't this list be something like:

```
            , result = [512, 256, 128, 64, 32, 16, 8, 4, 2, 1, 0]
```

Don't all those values satisfy the condition (>= 0)?

jonathanknowles replied:
> Shouldn't this list be something like:
> ```
>             , result = [512, 256, 128, 64, 32, 16, 8, 4, 2, 1, 0]
> ```

Not quite!

At each step, the `shrinkWhileSteps` function yields the smallest matching shrink that it can find.

You can see this by comparing the following two examples:

## Example 1

In this sequence, the smallest value produced by `shrink` just happens to match the test condition:
```hs
>>> shrink 1024
[0,512,768,896,960,992,1008,1016,1020,1022,1023]

>>> shrinkWhileSteps (>= 0) shrink 1024
[0]
```

Step-by-step analysis:
```hs
>>> find (>= 0) (shrink 1024)
Just 0
>>> find (>= 0) (shrink 0)
Nothing
```
Since `0` cannot be shrunk further, we terminate there.

## Example 2

In this sequence, the smallest value produced by `shrink` does not match the test condition, so we're forced to pick the next smallest value that does. This reveals the exponential back-off strategy of the `shrink @Int` function:

```hs
>>> shrink 1024
[0,512,768,896,960,992,1008,1016,1020,1022,1023]

>>> shrinkWhileSteps (>= 100) shrink 1024
[512,256,128,112,105,102,101,100]
```

Step-by-step analysis:
```hs
>>> find (>= 100) (shrink 1024)
Just 512
>>> find (>= 100) (shrink 512)
Just 256
>>> find (>= 100) (shrink 256)
Just 128
>>> find (>= 100) (shrink 128)
Just 112
>>> find (>= 100) (shrink 112)
Just 105
>>> find (>= 100) (shrink 105)
Just 102
>>> find (>= 100) (shrink 102)
Just 101
>>> find (>= 100) (shrink 101)
Just 100
>>> find (>= 100) (shrink 100)
Nothing
```

jonathanknowles replied:
One way to think about `shrinkWhileSteps` is that it generates a "slice" of the shrink space for a given value and shrinker, where all the values in that "slice" satisfy the condition.

It's designed to mimic the behaviour of QuickCheck when searching for a minimal counterexample that satisfies a given condition.

jonathanknowles replied:
@sevanspowell I've added a clarification to the comments for `shrinkWhile` and `shrinkWhileSteps` here: 302bf78d074f69dca45d9a25ffd8daf6e2f8599e

sevanspowell commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -208,6 +213,61 @@ shrinkInterleaved (a, shrinkA) (b, shrinkB) = interleave
 -- Evaluating shrinkers
 --------------------------------------------------------------------------------
 
+-- | Computes the entire shrink space of a given value and shrinking function.
```
It might be worth including a note to say that the shrink space of a given value and shrinking function is not necessarily equal to the entire value space of a type.

For example if the shrinking function for `Int` never chooses the value `7`, this function will also never produce the value `7`.

That became obvious to me after reading the implementation of this function and thinking about the type signature, but I don't think it's obvious on first glance.

jonathanknowles replied:
Hi @sevanspowell 

I've tried to clarify the comment for `shrinkSpace` a bit, and give some more worked examples.

See: ce310edacc949f1c751a4e7fcf04d6e1007221c7

jonathanknowles committed:
Clarify the utility of `shrinkWhile` and `shrinkWhileSteps`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3325#discussion_r896353185

jonathanknowles committed:
Clarify the behaviour of `shrinkSpace`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3325#discussion_r896361497

jonathanknowles committed:
Add property `prop_shrinkSpace_empty`.

jonathanknowles committed:
Add property `prop_shrinkSpace_singleton`.
